### PR TITLE
Refactor exporters to accept SpanData objects instead of Spans

### LINKED ISSUE #127: Pass raw spans to exporters
Currently, span objects are converted to an intermediate json format before being sent to an exporter. As a result, there are some stackdriver specific conversions [being applied](https://github.com/census-instrumentation/opencensus-python/blob/master/opencensus/trace/span.py#L222) to all spans (such as string truncation) that may not necessarily apply to all export formats.

I am proposing to update the [base.Exporter](https://github.com/census-instrumentation/opencensus-python/blob/master/opencensus/trace/exporters/base.py#L18) emit and export methods to take a raw [span.Span](https://github.com/census-instrumentation/opencensus-python/blob/master/opencensus/trace/span.py#L28) object instead of a dict. Not only will this push stackdriver specific details to the stackdriver exporter, but it would also make the development of new exporters easier since there are less layers of conversion.

Happy to make a PR if everyone is ok with the change.

### PR DESCRIPTION
**Pull request by wkiser**
Fixes #127

### CODE REVIEW

**Comment by wkiser** [comment]
@liyanhui1228, could you kick off circle again? It looks like it's stuck.

File: opencensus/trace/exporters/stackdriver_exporter.py [hunk 1]
```diff
@@ -132,33 +132,46 @@ def __init__(self, client=None, project_id=None,
...
+        :type span_datas: list of :class:
+            `~opencensus.trace.span_data.SpanData`
+        :param list of opencensus.trace.span_data.SpanData span_datas:
+            SpanData tuples to emit
         """
         name = 'projects/{}'.format(self.project_id)
-        stackdriver_spans = self.translate_to_stackdriver(spans)
+
+        # convert to the legacy trace json for easier refactoring
+        # todo refactor this to use the span data directly
```
Comment (liyanhui1228) [inline opencensus/trace/exporters/stackdriver_exporter.py]: Please change the `todo` to uppercase `TODO:`.

File: opencensus/trace/exporters/file_exporter.py [hunk 2]
```diff
@@ -47,14 +48,25 @@ def __init__(self, file_name=DEFAULT_FILENAME,
...
-        :param trace: Trace collected.
+        :type span_datas: list of :class:
+            `~opencensus.trace.span_data.SpanData`
+        :param list of opencensus.trace.span_data.SpanData span_datas:
+            SpanData tuples to emit
         """
         with open(self.file_name, self.file_mode) as file:
-            trace_str = json.dumps(trace)
+            # convert to the legacy trace json for easier refactoring
+            # todo refactor this to use the span data directly
```
Comment (liyanhui1228) [inline opencensus/trace/exporters/file_exporter.py]: Please change the `todo` to uppercase `TODO:`.

File: opencensus/trace/span_data.py [hunk 3]
```diff
@@ -81,6 +89,81 @@ class SpanData(_SpanData):
...
+    :return: Legacy 'trace' dictionary representing given SpanData tuples
+    """
+    if not span_datas:
+        return {}
+    top_span = span_datas[0]
+    assert isinstance(top_span, SpanData)
+    trace_id = None
+    if top_span.context is not None:
+        trace_id = top_span.context.trace_id
+    return {
```
Comment (liyanhui1228) [inline opencensus/trace/span_data.py]: assert the `trace_id` is not None before return?

File: opencensus/trace/span_data.py [hunk 4]
```diff
@@ -81,6 +89,81 @@ class SpanData(_SpanData):
...
+    :rtype: dict
+    :return: Legacy 'trace' dictionary representing given SpanData tuples
+    """
+    if not span_datas:
+        return {}
+    top_span = span_datas[0]
+    assert isinstance(top_span, SpanData)
+    trace_id = None
+    if top_span.context is not None:
+        trace_id = top_span.context.trace_id
```
Comment (liyanhui1228) [inline opencensus/trace/span_data.py]: Could be simplified to ` trace_id = top_span.context.trace_id if top_span.context is not None else None`.

**Comment by liyanhui1228** [comment]
We need to disable circleci running the default nosetests, which runs the unittest before installing the test requirements. And override it with our nox tests command. I'm preparing the PR for updating the circleci config.

**Pull request merged**